plugins/methodinfo/src/main/java/org/qbicc/plugin/methodinfo/MethodDataEmitter.java

```diff
@@ -0,0 +1,382 @@
+        ValueType uint32Type = ts.getUnsignedInteger32Type();
+        CompoundType sourceCodeInfoType = CompoundType.builder(ts)
+            .setTag(CompoundType.Tag.STRUCT)
+            .setName("qbicc_souce_code_info")
+            .setOverallAlignment(uint32Type.getAlign())
+            .addNextMember("methodInfoIndex", uint32Type)
+            .addNextMember("lineNumber", uint32Type)
+            .addNextMember("bcIndex", uint32Type)
+            .addNextMember("inlinedAtIndex", uint32Type)
+            .build();
+
+        Literal[] scInfoLiterals = Arrays.stream(scList).parallel().map(scInfo -> {
+            HashMap<CompoundType.Member, Literal> valueMap = new HashMap<>();
+            valueMap.put(sourceCodeInfoType.getMember(0), lf.literalOf(scInfo.getMethodInfoIndex()));
+            valueMap.put(sourceCodeInfoType.getMember(1), lf.literalOf(scInfo.getLineNumber()));
+            valueMap.put(sourceCodeInfoType.getMember(2), lf.literalOf(scInfo.getBcIndex()));
+            valueMap.put(sourceCodeInfoType.getMember(3), lf.literalOf(scInfo.getInlinedAtIndex()));
+            return lf.literalOf(sourceCodeInfoType, valueMap);
+        }).toArray(Literal[]::new);
+
+        return lf.literalOf(ts.getArrayType(sourceCodeInfoType, scInfoLiterals.length), List.of(scInfoLiterals));
+    }
+
+    Literal emitInstructionMap(CompilationContext ctxt, InstructionMap[] imapList) {
+        TypeSystem ts = ctxt.getTypeSystem();
+        LiteralFactory lf = ctxt.getLiteralFactory();
+        Section section = ctxt.getImplicitSection(ctxt.getDefaultTypeDefinition());
+        ValueType uint64Type = ts.getUnsignedInteger64Type();
+        CompoundType imapType = CompoundType.builder(ts)
+            .setTag(CompoundType.Tag.STRUCT)
+            .setName("qbicc_instruction_map")
+            .setOverallAlignment(uint64Type.getAlign())
+            .addNextMember("instructionAddress", uint64Type)
+            .addNextMember("sourceCodeIndex", uint64Type)
+            .build();
+
+        Literal[] imapLiterals = IntStream.range(0, imapList.length)
+            .parallel()
+            .mapToObj(i -> {
+            HashMap<CompoundType.Member, Literal> valueMap = new HashMap<>();
+            Function function = ctxt.getExactFunction(imapList[i].getFunction());
+            Literal functionCastLiteral = lf.bitcastLiteral(lf.literalOfSymbol(function.getName(), function.getType().getPointer()), ts.getUnsignedInteger8Type().getPointer());
+            Literal instructionAddrLiteral = lf.valueConvertLiteral(lf.elementOfLiteral(functionCastLiteral, lf.literalOf(imapList[i].getOffset())), ts.getUnsignedInteger64Type());
+            valueMap.put(imapType.getMember(0), instructionAddrLiteral);
+            valueMap.put(imapType.getMember(1), lf.literalOf(imapList[i].getSourceCodeIndex()));
+            section.declareFunction(null, function.getName(), function.getType());
+            return lf.literalOf(imapType, valueMap);
+        }).toArray(Literal[]::new);
```

Comment:
Is this generating full instruction addresses?  Do appropriate relocations get generated by LLVM for this?

Reply:
Yes, the instruction map table in the final binary contains the full instruction address.